### ISSUE TITLE
About GatherLayer backward

### ISSUE THREAD

**Issue by Pexure** [open-mmlab/mmselfsup#107]
Hi there. Thanks for providing the implementation of GatherLayer forward and backward. But as for backward, what's the point of creating new zero tensors and assign them the grads? Can I just simply ```return grads[dist.get_rank()]``` for backward()?
```
class GatherLayer(torch.autograd.Function):
    """Gather tensors from all process, supporting backward propagation.
    """

    @staticmethod
    def forward(ctx, input):
        ctx.save_for_backward(input)
        output = [torch.zeros_like(input) \
            for _ in range(dist.get_world_size())]
        dist.all_gather(output, input)
        return tuple(output)

    @staticmethod
    def backward(ctx, *grads):
        input, = ctx.saved_tensors
        grad_out = torch.zeros_like(input)
        grad_out[:] = grads[dist.get_rank()]
        return grad_out
```

**Comment by XiaohangZhan** [comment]
I think you'd better not. Variable `grads` is the gradient of the output, here we create a new variable `grad_out` for the input of this layer. If the gradient of the input and the output point to the same memory area, then any modifications to some gradient might affect other gradients which may not be expected.

**Comment by Pexure** [comment]
Got it. Thanks
